Kitodo/src/main/java/org/kitodo/production/helper/tasks/HierarchyMigrationTask.java

```diff
@@ -0,0 +1,358 @@
+/*
+ * (c) Kitodo. Key to digital objects e. V. <contact@kitodo.org>
+ *
+ * This file is part of the Kitodo project.
+ *
+ * It is licensed under GNU General Public License version 3 or later.
+ *
+ * For the full copyright and license information, please read the
+ * GPL3-License.txt file that was distributed with this source code.
+ */
+
+package org.kitodo.production.helper.tasks;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kitodo.api.MetadataEntry;
+import org.kitodo.api.dataformat.IncludedStructuralElement;
+import org.kitodo.api.dataformat.Workpiece;
+import org.kitodo.api.dataformat.mets.LinkedMetsResource;
+import org.kitodo.data.database.beans.Process;
+import org.kitodo.data.database.beans.Project;
+import org.kitodo.data.database.exceptions.DAOException;
+import org.kitodo.data.exceptions.DataException;
+import org.kitodo.exceptions.ProcessGenerationException;
+import org.kitodo.production.helper.Helper;
+import org.kitodo.production.metadata.MetadataEditor;
+import org.kitodo.production.process.ProcessGenerator;
+import org.kitodo.production.services.ServiceManager;
+import org.kitodo.production.services.data.ProcessService;
+import org.kitodo.production.services.dataeditor.DataEditorService;
+import org.kitodo.production.services.dataformat.MetsService;
+import org.kitodo.production.services.file.FileService;
+
+public class HierarchyMigrationTask extends EmptyTask {
+    private static final Logger logger = LogManager.getLogger(HierarchyMigrationTask.class);
+
+    /**
+     * Service that contains the meta-data editor.
+     */
+    private static final DataEditorService dataEditorService = ServiceManager.getDataEditorService();
+
+    /**
+     * Service to access files on the storage.
+     */
+    private static final FileService fileService = ServiceManager.getFileService();
+
+    /**
+     * Service to read and write METS file format.
+     */
+    private static final MetsService metsService = ServiceManager.getMetsService();
+
+    /**
+     * Service to generate processes.
+     */
+    private final ProcessGenerator processGenerator = new ProcessGenerator();
+
+    /**
+     * Service to read and write Process objects in the database or search
+     * engine index.
+     */
+    private static final ProcessService processService = ServiceManager.getProcessService();
+
+    /**
+     * Parent processes have already been created. Key is the identifier, the
+     * value is the process ID, then the current numbers of child links.
```

Comment:
You are mixing up purposes of integers in the List.
Why not using the process ID as a key instead of the String?
Shouldn't the list of childs also be found in the Database?

Reply:
* I need several integers to do this job. These integer values are saved in the list for each parent process.
* I do not use the ID as a key, because in further child processes I can only find the string as the parent, but not the ID. So with the ID as a key, I cannot check whether it has already created the parent or not. This is basically what the map is for.
* The list of children can be found in the database and is not necessary here. The list contains the current number of the children, not their object ID.

Reply:
"because in further child processes I can only find the string as the parent"  What does that mean? The string of the ID? Or what string?
Is it correct, that the list always only contains two entries? parentID and number of processes? maybe use a Map?

Reply:
The string is the PPN of the superior unit (the journal, or the multi-volume works as its whole). It is found in the anchor.xml file as `CatalogIDDigital`. This is used to identify the shared parent of several processes. First, this string is read from the file. Then the migration looks if this string is in the map. If yes, then the parent process has been created (there was already another child for this parent before). If not, a parent process is created and written into the map.

The number of integers in the list is always one greater as the number of issues already found for the parent, because first there is the parent record number. So it is two after one child, but more after more children.

**An example:**
- The migration finds a hierarchial process (database no. 22) and reads parent PPN `1234567X` from the anchor file, and `CurrentNo` 2 from the meta.xml file. The map is empty. So it creates a new process as parent process. That process gets database number 46. Then it writes the following entry into the map: `"1234567X" → <46, 2>`. It adds a link in the logical tree from the parent to the child no. 22.
- Then it finds a hierarchical process (database no. 35) and reads parent PPN `1234567X` from the anchor file, and `CurrentNo` 3 from the meta.xml file. There is an entry for string `1234567X` in the map, so there is already a parent for this and no new process is created. It adds a link in the logical tree from the parent to the child no. 35 _after_ the first link, because it knows (from the int list) that the first link is current no. 2 and this is current no. 3 and must therefore be added after the first link. Then it updates the map entry to look like this: `"1234567X" → <46, 2, 3>`
- Then it finds a hierarchical process (database no. 19) and reads parent PPN `23456790` from the anchor file, and `CurrentNo` 3 from the meta.xml file. There is no entry for string `23456790` in the map, so it creates a new process as parent process. That process gets database number 47. Then it writes the following entry into the map: `"23456790" → <47, 3>`. It adds a link in the logical tree from the parent to the child no. 19.
- Then it finds a hierarchical process (database no. 17) and reads parent PPN `23456790` from the anchor file, and `CurrentNo` 1 from the meta.xml file. There is an entry for string `23456790` in the map, so there is already a parent for this and no new process is created. It adds a link in the logical tree from the parent to the child no. 17 _before_ the first link, because it knows (from the int list) that the first link is current no. 3 and this is current no. 1 and must therefore be added before the first link. Then it updates the map entry to look like this: `"23456790" → <47, 1, 3>`
- Then it finds a hierarchical process (database no. 45) and reads parent PPN `1234567X` from the anchor file, and `CurrentNo` 5 from the meta.xml file. There is an entry for string `1234567X` in the map, so there is already a parent for this and no new process is created. It adds a link in the logical tree from the parent to the child no. 45 _after_ the two other links, because it knows (from the int list) that the first link is current no. 2 and the second link is current no. 3 and this is current no. 5 and must therefore be added after the two links. Then it updates the map entry to look like this: `"1234567X" → <46, 2, 3, 5>`
- Then it finds a hierarchical process (database no. 18) and reads parent PPN `23456790` from the anchor file, and `CurrentNo` 2 from the meta.xml file. There is an entry for string `23456790` in the map, so there is already a parent for this and no new process is created. It adds a link in the logical tree from the parent to the child no. 18 _in between_ the first and second link, because it knows from the int list that the first link is current no. 1 and the second link is current no. 3 and this is current no. 2 and must therefore be added between the two links. Then it updates the map entry to look like this: `"23456790" → <47, 1, 2, 3>`
- Then it finds a hierarchical process (database no. 40) and reads parent PPN `1234567X` from the anchor file, and `CurrentNo` 4 from the meta.xml file. There is an entry for string `1234567X` in the map, so there is already a parent for this and no new process is created. It adds a link in the logical tree from the parent to the child no. 40 between the second and third link, because it knows from the int list that the second link is current no. 3 and the third link is current no. 5 and this is current no. 4 and must therefore be added between the second and third link. Then it updates the map entry to look like this: `"1234567X" → <46, 2, 3, 4, 5>`
- Then it finds a hierarchical process (database no. 14) and reads parent PPN `1234567X` from the anchor file, and `CurrentNo` 1 from the meta.xml file. There is an entry for string `1234567X` in the map, which looks like this: `"1234567X" → <46, 2, 3, 4, 5>`. Should be clear now where to add the link …